File: debug.go
package groundcover

import (
	"fmt"
	"sort"
	"strings"
)

// debugMaxFrames bounds how many stack frames the debug renderer prints.
const debugMaxFrames = 8

// debugMaxValueLen bounds the rendered length of an attribute value.
const debugMaxValueLen = 80

// renderDebug formats a finalized event as a compact, human-readable block for
// local development. It renders the post-scrub, post-hash event, so it honors
// BeforeSend scrubbing and IdentityHasher pseudonymization.
func renderDebug(e *Event) string {
	var b strings.Builder
	fmt.Fprintf(&b, "[groundcover] %s %s  %s\n", e.Level, e.Type, e.Title)
	fmt.Fprintf(&b, "  fingerprint=%s handled=%t\n", e.Fingerprint, e.ErrorHandled)

	if !e.User.isZero() {
		fmt.Fprintf(&b, "  user: %s\n", userSummary(e.User))
	}
	if len(e.Attributes) > 0 {
		fmt.Fprintf(&b, "  attrs: %s\n", attrsSummary(e.Attributes))
	}
	if n := len(e.Stacktrace); n > 0 {
		b.WriteString("  stack:\n")
		shown := min(n, debugMaxFrames)
		for _, f := range e.Stacktrace[:shown] {
			fmt.Fprintf(&b, "    %s (%s:%d)\n", f.Function, f.File, f.Line)
		}
		if n > shown {
			fmt.Fprintf(&b, "    ... %d more\n", n-shown)
		}
	}
	return b.String()
}

func userSummary(u User) string {
	parts := make([]string, 0, 4)
	if u.ID != "" {
		parts = append(parts, "id="+u.ID)
	}
	if u.Email != "" {
		parts = append(parts, "email="+u.Email)
	}
	if u.Name != "" {
		parts = append(parts, "name="+u.Name)
	}
	if u.Organization != "" {
		parts = append(parts, "org="+u.Organization)
	}
	return strings.Join(parts, " ")
}

func attrsSummary(a Attributes) string {
	keys := make([]string, 0, len(a))
	for k := range a {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	parts := make([]string, 0, len(keys))
	for _, k := range keys {
		parts = append(parts, k+"="+truncateValue(fmt.Sprintf("%v", a[k])))
	}
	return strings.Join(parts, " ")
}

func truncateValue(s string) string {
	if len(s) <= debugMaxValueLen {
		return s
	}
	return s[:debugMaxValueLen-1] + "\u2026"
}

File: groundcover-python/src/groundcover/_debug.py
"""Compact human-readable event rendering for local development."""

from __future__ import annotations

from ._event import Attributes, Event
from ._user import User

DEBUG_MAX_FRAMES = 8
"""Bounds how many stack frames the debug renderer prints."""

DEBUG_MAX_VALUE_LEN = 80
"""Bounds the rendered length of an attribute value."""


def render_debug(e: Event) -> str:
    """Format a finalized event as a compact, human-readable block for local
    development. It renders the post-scrub, post-hash event, so it honors
    before_send scrubbing and IdentityHasher pseudonymization."""
    lines = [
        f"[groundcover] {e.level.value} {e.type}  {e.title}",
        f"  fingerprint={e.fingerprint} handled={'true' if e.error_handled else 'false'}",
    ]
    if not e.user.is_zero():
        lines.append(f"  user: {_user_summary(e.user)}")
    if e.attributes:
        lines.append(f"  attrs: {_attrs_summary(e.attributes)}")
    if e.stacktrace:
        lines.append("  stack:")
        shown = min(len(e.stacktrace), DEBUG_MAX_FRAMES)
        for f in e.stacktrace[:shown]:
            lines.append(f"    {f.function} ({f.file}:{f.line})")
        if len(e.stacktrace) > shown:
            lines.append(f"    ... {len(e.stacktrace) - shown} more")
    return "\n".join(lines) + "\n"


def _user_summary(u: User) -> str:
    parts = []
    if u.id:
        parts.append(f"id={u.id}")
    if u.email:
        parts.append(f"email={u.email}")
    if u.name:
        parts.append(f"name={u.name}")
    if u.organization:
        parts.append(f"org={u.organization}")
    return " ".join(parts)


def _attrs_summary(a: Attributes) -> str:
    parts = []
    for k in sorted(a):
        parts.append(f"{k}={_truncate_value(str(a[k]))}")
    return " ".join(parts)


def _truncate_value(s: str) -> str:
    if len(s) <= DEBUG_MAX_VALUE_LEN:
        return s
    return s[: DEBUG_MAX_VALUE_LEN - 1] + "\u2026"
